Src/Base/AMReX_TinyProfiler.cpp

```diff
@@ -134,6 +141,225 @@ TinyProfiler::stop () noexcept
         stats.clear();
     }
 }
+#else
+TinyProfiler::TinyProfiler (std::string funcname) noexcept
+    : fname(std::move(funcname)), uCUPTI(false)
+{
+    start();
+}
+
+TinyProfiler::TinyProfiler (std::string funcname, bool start_) noexcept
+    : fname(std::move(funcname)), uCUPTI(false)
+{
+    if (start_) start();
+}
+
+TinyProfiler::TinyProfiler (std::string funcname, bool start_, bool useCUPTI) noexcept
+    : fname(std::move(funcname)), uCUPTI(useCUPTI)
+{
+    if (start_) start();
+}
+
+TinyProfiler::TinyProfiler (const char* funcname) noexcept
+    : fname(funcname), uCUPTI(false)
+{
+    start();
+}
+
+TinyProfiler::TinyProfiler (const char* funcname, bool start_) noexcept
+    : fname(funcname), uCUPTI(false)
+{
+    if (start_) start();
+}
+
+TinyProfiler::TinyProfiler (const char* funcname, bool start_, bool useCUPTI) noexcept
+    : fname(funcname), uCUPTI(useCUPTI)
+{
+    if (start_) start();
+}
+
+TinyProfiler::~TinyProfiler ()
+{
+    stop();
+}
+
+void
+TinyProfiler::start () noexcept
+{
+#ifdef _OPENMP
+#pragma omp master
+#endif
+    if (stats.empty() && !regionstack.empty())
+    {
+        double t;
+        if (!uCUPTI) {
+	  t = amrex::second();
+	} else {
+	  cudaDeviceSynchronize();
+	  cuptiActivityFlushAll(0);
+	  activityRecordUserdata.clear();
+	  t = amrex::second();
+	}
+	
+	ttstack.emplace_back(std::make_tuple(t, 0.0, &fname));
+	global_depth = ttstack.size();
+
+#ifdef AMREX_USE_CUDA
+	nvtx_id = nvtxRangeStartA(fname.c_str());
+#endif
+
+        for (auto const& region : regionstack)
+        {
+            Stats& st = statsmap[region][fname];
+            ++st.depth;
+            stats.push_back(&st);
+        }
+    }
+}
+
+void
+TinyProfiler::stop () noexcept
+{
+#ifdef _OPENMP
+#pragma omp master
+#endif
+    if (!stats.empty()) 
+    {
+        double t;
+        int nKernelCalls = 0;
+        	
+        if (!uCUPTI) {
+	  t = amrex::second();
+	} else {
+	  cudaDeviceSynchronize();
+	  cuptiActivityFlushAll(0);
+	  t = computeElapsedTimeUserdata(activityRecordUserdata);
+	  nKernelCalls = activityRecordUserdata.size();
+	}
+
```

Comment:
cuptiActivityFlushAll is called at every start and stop? Is that required?

Reply:
As is, this is how we are able to tag records with a loop iteration marker.  It is not ideal, so we can think about a better way to handle this.